### DISCUSSION TITLE
Flowfield visualization doubts

### DISCUSSION THREAD

**Discussion by ghost** [NatLabRockies/floris#455] [Q&A] [answered]
Hello,

I have doubts regarding the visualization of the flow field.
We have tried to plot the flow field for different wind directions (same layout) and compare it with the original layout.
However, the positions of the "turbines" in the flow field do not correspond to those in the original layout (except for the case wind_dir=270º).
Interpreting the image with reference systems with respect to the wind, I haven't been able to understand it either, because the x and y axes weren't interchanged when turning it 90º.
Sorry if the question is too basic

1. Case wd =0
![wd_0](https://user-images.githubusercontent.com/98586208/176180967-02649f82-fac5-4b21-95d4-dc7bcabd1aa1.png)
2. Case wd =90
![wd_90](https://user-images.githubusercontent.com/98586208/176180957-9216e6eb-7808-407d-9615-d6f383c952c6.png)
3. Case wd =270
![wd_270](https://user-images.githubusercontent.com/98586208/176180972-08054a2d-185e-4438-969d-485315b535f6.png)
4. Case wd =180
![wd_180](https://user-images.githubusercontent.com/98586208/176180976-eb3c7451-1b85-44cf-9018-1306d26fd196.png)
```

from floris.tools import FlorisInterface
import matplotlib.pyplot as plt
from floris.tools.visualization import visualize_cut_plane
fi = FlorisInterface('gch.yaml')
layout_x = [0,0,300,300,900,900]
layout_y= [10000,10400,10000,10400,10000,10400]

fi.reinitialize(
    layout=[layout_x, layout_y],
    wind_directions=[0],
    wind_speeds=[8]
)
fig, ax = plt.subplots()
horizontal_plane = fi.calculate_horizontal_plane(height=90.0)
visualize_cut_plane(horizontal_plane, ax=ax, title="Horizontal, wind_dir=0")
ax.scatter(layout_x, layout_y, c='k')`
```

**Comment by rafmudaf** [comment]
Are the plots correct with respect to their labels? See the title on the plot and the label in the pull request form. 3 and 4 may be reversed.

**Reply by ghost** [reply]
You are right. i have corrected the comment.

**Comment by rafmudaf** [comment, marked as answer]
@mapasan The plot always shows the wind coming from the left and rotates the turbine layout accordingly. For example, here's the plot from your script above with the wind direction at 275

![Figure_1](https://user-images.githubusercontent.com/13797903/176233571-45f12a08-5d2e-4e9a-8e7b-6f0b4fb79de6.png)

You can see that the black dots are in their unrotated locations and the turbines that are shown in the velocity profile are slightly rotated about the midpoint of the whole farm. In other words, we keep the wind direction coming from the left on the plot and rotate the farm, as needed; see [rotate_coordinates_rel_west](https://github.com/NREL/floris/blob/main/floris/utilities.py#L196).

Does this answer your question?

**Reply by ghost** [reply]
Thank you Rafael! Now I understand the figures!
However, for me the ticks on the graphs are quite confusing as they do not represent the absolute coordinates in the "original" reference system, and the identification of the "id" of the turbines might not be obvious. A possible improvement could be the addition of "ids" to the figure.
Do you expect to develop graphics in the original coordinate reference system?
Thank you very much!! :)

**Reply by rafmudaf** [reply]
I completely agree @mapasan. I think we are accustomed to looking at these plots, but they are indeed confusing.

Plotting the unrotated farm (or "original" reference frame) is difficult because the FLORIS domain is a rectangle oriented with the wind direction. When we plot the original coordinates, matplotlib displays the FLORIS domain (the rectangle wrt wind direction) inside of the plotting window which is a rectangle oriented with the computer screen. The result is a mostly white (empty) figure with the FLORIS domain partially fit into the window. I hope this makes sense, and I will try to make an example.

Truthfully, we have put just enough effort into our plotting routines to get something that we hope is _good enough_ for many use cases. That being said, this is an area that I'd love to continue to improve, and its a great place for community contribution. A good starting point is to outline specific feature ideas and requirements. Could you start a new discussion thread on visualization and include your suggestions? I will add a few of my own, as well.
